### PR TITLE
[PyVelox] Add type specification support while creating vectors from list

### PR DESCRIPTION
**Pull request by sanjibansg**
This PR adds the support to specify data type while creating a vector from a Python list. With this change, the user can also create empty or null vectors.

### CODE REVIEW

File: pyvelox/test/test_vector.py [hunk 1]
```diff
@@ -55,6 +55,17 @@ def test_from_list(self):
         with self.assertRaises(ValueError):
             pv.from_list([])
 
+    def test_from_list_with_type(self):
+        a = pv.from_list([0, 1, 3], pv.BooleanType())
+        self.assertEqual(a.typeKind().name, 'BOOLEAN')
+        for i in range(len(a)):
+            self.assertTrue(isinstance(a[i], bool))
```
[the remark this answers is not in the document]
Comment (kgpai) [reply pyvelox/test/test_vector.py:62]: Can we add a test to convert say a number to varchar and test when conversions cannot happen etc. 
Also test cases when conversions happen with precision loss etc.

[the remark this answers is not in the document]
Comment (sanjibansg) [reply pyvelox/test/test_vector.py:63]: Added the tests for:

- Checking with the boolean value while `int` to `bool` conversion
- Checking conversion is not possible while casting `int` to `varchar`
- Checking for precision loss

File: pyvelox/test/test_vector.py [hunk 2]
```diff
@@ -55,6 +55,31 @@ def test_from_list(self):
         with self.assertRaises(ValueError):
             pv.from_list([])
 
+    def test_from_list_with_type(self):
+        list_a = [0, 1, 3]
+        a = pv.from_list(list_a, pv.BooleanType())
```
Comment (sanjibansg) [inline pyvelox/test/test_vector.py:60]: Does the function signature look good? Or will it be better if the data type is specified by a keyword argument?

Ref: Numpy array creation by data type: https://numpy.org/doc/stable/user/basics.creation.html
Torch tensor creation by data type: https://pytorch.org/docs/stable/tensors.html#initializing-and-basic-operations

File: pyvelox/pyvelox.h [hunk 3]
```diff
@@ -69,6 +69,50 @@ inline velox::variant pyToVariant(const py::handle& obj) {
   }
 }
 
+inline velox::variant pyToVariant(const py::handle& obj, const Type& dtype) {
+  if (obj.is_none()) {
+    return velox::variant();
```
Comment (kgpai) [inline pyvelox/pyvelox.h:74]: Even if obj is none, should we loose all type information ?

Comment (kgpai) [reply pyvelox/pyvelox.h:74]: We can create a null variant which has typekind information, we should use that imo.

File: pyvelox/test/test_vector.py (hunk 2 above)
Comment (kgpai) [reply pyvelox/test/test_vector.py:60]: Its an interesting semantic choice you bring up - would it be possible to support both position and keyword  ?

Comment (sanjibansg) [reply pyvelox/test/test_vector.py:60]: Made the change. It now supports passing the data type both as a positional and a keyword argument. Right now, the default value is None, and if it is not passed, the data type is inferred.

File: pyvelox/pyvelox.h [hunk 4]
```diff
@@ -69,6 +69,50 @@ inline velox::variant pyToVariant(const py::handle& obj) {
...
+  } else {
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply pyvelox/pyvelox.h]: Thanks for pointing that out. Removed it.

File: pyvelox/pyvelox.h (hunk 3 above)
Comment (sanjibansg) [reply pyvelox/pyvelox.h]: Tried returning a null variant with type information. Does this look good?

**Review by kgpai — APPROVED** [review]

**Comment by kgpai** [comment]
@sanjibansg Can you fix failing pyvelox build due to this PR ?

**Pull request closed without merging**